### ISSUE TITLE
Open / Save Dialog - Max filepath bug limit

### ISSUE THREAD

**Issue by Coloris** [spiegelp/MaterialDesignExtensions#22]
Trying to enter a very long filename result of a break :

> System.IO.PathTooLongException
  HResult=0x800700CE
  Message=The specified path, file name, or both are too long. The fully qualified file name must be less than 260 characters, and the directory name must be less than 248 characters.
  Source=mscorlib
  StackTrace:
   at System.IO.Path.LegacyNormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
   at System.IO.Path.NormalizePath(String path, Boolean fullCheck, Int32 maxPathLength, Boolean expandShortPaths)
   at System.IO.Path.GetFullPathInternal(String path)
   at System.IO.FileInfo.Init(String fileName, Boolean checkHost)
   at System.IO.FileInfo..ctor(String fileName)
   at MaterialDesignExtensions.Controls.FileSelectedEventArgs..ctor(RoutedEvent routedEvent, Object source, String file)
   at MaterialDesignExtensions.Controls.BaseFileControl.SelectFileCommandHandler(Object sender, ExecutedRoutedEventArgs args)
   at MaterialDesignExtensions.Controls.SaveFileControl.SelectFileCommandHandler(Object sender, ExecutedRoutedEventArgs args)
   at System.Windows.Input.CommandBinding.OnExecuted(Object sender, ExecutedRoutedEventArgs e)
   at System.Windows.Input.CommandManager.ExecuteCommandBinding(Object sender, ExecutedRoutedEventArgs e, CommandBinding commandBinding)
   at System.Windows.Input.CommandManager.FindCommandBinding(CommandBindingCollection commandBindings, Object sender, RoutedEventArgs e, ICommand command, Boolean execute)
   at System.Windows.Input.CommandManager.FindCommandBinding(Object sender, RoutedEventArgs e, ICommand command, Boolean execute)
   at System.Windows.Input.CommandManager.OnExecuted(Object sender, ExecutedRoutedEventArgs e)
   at System.Windows.UIElement.OnExecutedThunk(Object sender, ExecutedRoutedEventArgs e)
   at System.Windows.Input.ExecutedRoutedEventArgs.InvokeEventHandler(Delegate genericHandler, Object target)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs args, Boolean trusted)
   at System.Windows.Input.RoutedCommand.ExecuteImpl(Object parameter, IInputElement target, Boolean userInitiated)
   at System.Windows.Input.RoutedCommand.ExecuteCore(Object parameter, IInputElement target, Boolean userInitiated)
   at MS.Internal.Commands.CommandHelpers.CriticalExecuteCommandSource(ICommandSource commandSource, Boolean userInitiated)
   at System.Windows.Controls.Primitives.ButtonBase.OnClick()
   at System.Windows.Controls.Button.OnClick()
   at System.Windows.Controls.Primitives.ButtonBase.OnMouseLeftButtonUp(MouseButtonEventArgs e)
   at System.Windows.UIElement.OnMouseLeftButtonUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.ReRaiseEventAs(DependencyObject sender, RoutedEventArgs args, RoutedEvent newEvent)
   at System.Windows.UIElement.OnMouseUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(Delegate genericHandler, Object genericTarget)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs args, Boolean trusted)
   at System.Windows.Input.InputManager.ProcessStagingArea()
   at System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   at System.Windows.Input.InputProviderSite.ReportInput(InputReport inputReport)
   at System.Windows.Interop.HwndMouseInputProvider.ReportInput(IntPtr hwnd, InputMode mode, Int32 timestamp, RawMouseActions actions, Int32 x, Int32 y, Int32 wheel)
   at System.Windows.Interop.HwndMouseInputProvider.FilterMessage(IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Threading.Dispatcher.PushFrame(DispatcherFrame frame)
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at System.Windows.Application.Run(Window window)
   at System.Windows.Application.Run()
   at AppTest.App.Main()

Maybe it would be possible to evaluate if the path is too long before validating it ? Another option would be to use unicode the same way as AlphaFS does.

**Comment by spiegelp** [comment]
@Coloris On the one hand implemting native support for long paths would be to much, error prone work. On the other hand I do not like to switch to AlphaFS, because it will break the public API. However .NET 4.6.2 seems to support long paths without any code modifications on Windows 10. 

As a quick and easy solution I could add a second build targeting .NET 4.6.2. Which .NET framework and OS are you targeting with your app?

**Comment by Coloris** [comment]
@spiegelp Well I totally understand your point but the thing is that not everyone has Windows 10 yet (and not all version of Windows 10 works with longpfilepath even).

Wouldn't it be possible to at least circumvent it in someway because I cannot even try catch the exception on my side... my app just break (like you could raise an exception yourself ?).

In my case I plan to support starting from Windows7 up to 10.

**Comment by spiegelp** [comment]
@Coloris I am thinking of creating a second build and nuget package using AlphaFS instead of `System.IO`. The new package would have the same API except of using `FileInfo` and `DirectoryInfo` from AlphaFS in order to support long paths on older .NET frameworks and operating systems. 

You will only need to reference the new nuget package. Will this solve the issue for your scenario?

**Comment by Siroga00VII** [comment]
private string BuildFullFilename(string newFilename)
        {
            string filename = null;

            if (!string.IsNullOrWhiteSpace(newFilename))
            {
                string directory = CurrentDirectory;

                if (CurrentDirectory != null && !directory.EndsWith(@"\") && !directory.EndsWith("/"))
                {
                    directory = directory + @"\";
                }

                var fInfo = new FileInfo(newFilename);


                var format = Filters[FilterIndex].Filters.TrimStart('*');


                if (format != ".*" && string.IsNullOrEmpty(fInfo.Extension))
                {
                    filename = directory + newFilename.Trim() + format;
                }
                else if (format != ".*" && fInfo.Extension != format)
                {
                    filename = directory + newFilename.Trim().Replace(fInfo.Extension, format);

                }
                else
                    filename = directory + newFilename.Trim();


            }

            return filename;
        }
        MaterialDesignExtensions-master\MaterialDesignExtensions\Controls\SaveFileControl.cs
        Error Please correct

**Comment by spiegelp** [comment]
I published the pre-release v2.7.0-a01 some moments ago. 

@Coloris After a long time of thinking of how to support long paths with minimal side effects, I published a new package [MaterialDesignExtensions.LongPath](https://www.nuget.org/packages/MaterialDesignExtensions.LongPath/). It has the same source code but uses Pri.LongPath instead of System.IO. This package should solve your issues on older Windows systems. 

@Siroga00VII Check out the newest pre-release. If you set `SaveFileDialog.ForceFileExtensionOfFileFilter = true`, the control enables the behavior to enforce file extensions.
